### ISSUE TITLE
Possible race condition on `get` operation

### ISSUE THREAD

**Issue by jochem-brouwer** [ethereumjs/merkle-patricia-tree#134]
The `get` operation of the trie does not have a lock. It seems to me that if you run a `get` (which calls `findPath`) and meanwhile also do a `put` operation, this would change the path (`get` operation runs on an old trie). If you would change the same key using `put` to a new value, then `get` will probably yield the old value. Not sure if this is a big problem though.

**Comment by zmitton** [comment]
I don't believe that constitutes a race condition because it only happens when the `get` and `put` are called in separate threads, and threads are already known to not have a guaranteed order.

It would, however, be a race condition if the `put` call, ended up deleting intermediary nodes, and then the `get` call got stuck on an intermediate node (non-leaf) as a missing node error.

**Comment by zmitton** [comment]
Note that if wrapping the same lock around `get` yields insufficient performance, you could instead have a separate lock instance that allows multiple threads accessing `get`, but locks out `put` and `delete` operations. The semephore package already supports something like this
